tcclient/creds.go

```diff
@@ -7,16 +7,75 @@ import (
 	"encoding/json"
 	"errors"
 	"fmt"
+	"net/http"
 	"strconv"
 	"strings"
 	"time"
 
 	"github.com/taskcluster/jsonschema2go/text"
 	"github.com/taskcluster/slugid-go/slugid"
+	"github.com/tent/hawk-go"
 )
 
-type Credentials struct {
-	ClientId    string `json:"clientId"`
+type Credentials interface {
+	fmt.Stringer
+	ConfigureAuth(httpRequest *http.Request) error
+	getExtField() (string, error)
+	HawkCredentials() *hawk.Credentials
```

Comment:
Thanks for all the feedback @jonasfj. In essence, we are aligned, I already had the upfront checking planned, and the refactoring of the interface - in fact I was nervous dumping work in progress as I anticipated you'd suggest exactly the parts I was about to do. :)

But it is good to see we are aligned, as these are pretty much the changes I was planning on making. The reason the interface was messy at the moment was because I just moved the existing functions into interface methods, and the next step was to clean them up.

I really prefer the new design so we can statically be sure that you cannot attempt to create temp credentials from temp credentials (unless you make an invalid cast, which would be your own fault), and that we get the upfront error checking, and also the immutability so we don't need to worry about concurrency. I also like that we can create a type to represent no credentials, and remove the boolean. Previously I had a boolean in there so you could temporarily disable authentication, without losing the credentials, but now that we've gone for immutable, that makes no sense.